channel/channelapi/state.go

```diff
@@ -0,0 +1,475 @@
+			}
+			if message.Embeds != nil {
+				m.Embeds = message.Embeds
+			}
+			if message.Attachments != nil {
+				m.Attachments = message.Attachments
+			}
+			if !message.Timestamp.IsZero() {
+				m.Timestamp = message.Timestamp
+			}
+			if message.Author != nil {
+				m.Author = message.Author
+			}
+			if message.Components != nil {
+				m.Components = message.Components
+			}
+
+			return nil
+		}
+	}
+
+	c.Messages = append(c.Messages, message)
+
+	if len(c.Messages) > s.GetMaxMessageCount() {
+		c.Messages = c.Messages[len(c.Messages)-s.GetMaxMessageCount():]
+	}
+
+	return nil
+}
+
+// MessageRemove removes a channel.Message from the current State.
+func (s *State) MessageRemove(message *channel.Message) error {
+	return s.MessageRemoveByID(message.ChannelID, message.ID)
+}
+
+// MessageRemoveByID removes a channel.Message by channelID and messageID from the current State.
+func (s *State) MessageRemoveByID(channelID, messageID string) error {
+	c, err := s.Channel(channelID)
+	if err != nil {
+		if errors.Is(err, state.ErrStateNotFound) {
+			return errors.Join(err, ErrChannelNotCached)
+		}
+		return err
+	}
+
+	s.GetMutex().Lock()
+	defer s.GetMutex().Unlock()
+
+	for i, m := range c.Messages {
+		if m.ID == messageID {
+			c.Messages = append(c.Messages[:i], c.Messages[i+1:]...)
+			return nil
+		}
+	}
+
+	return state.ErrStateNotFound
+}
+
+// Message gets a message by channel and message ID.
+func (s *State) Message(channelID, messageID string) (*channel.Message, error) {
+	c, err := s.Channel(channelID)
+	if err != nil {
+		if errors.Is(err, state.ErrStateNotFound) {
+			return nil, errors.Join(err, ErrChannelNotCached)
+		}
+		return nil, err
+	}
+
+	s.GetMutex().RLock()
+	defer s.GetMutex().RUnlock()
+
+	for _, m := range c.Messages {
+		if m.ID == messageID {
+			return m, nil
+		}
+	}
+
+	return nil, state.ErrStateNotFound
+}
+
+// ThreadListSync syncs guild threads with provided ones.
+// TODO: use gokord.ThreadListSync when event will be remade
+func (s *State) ThreadListSync(guildID string, channelIDs []string, threads []*channel.Channel, members []*channel.ThreadMember) error {
+	g, err := s.GuildState().Guild(guildID)
+	if err != nil {
+		if errors.Is(err, state.ErrStateNotFound) {
+			return errors.Join(err, ErrGuildNotCached)
+		}
+		return err
+	}
+
+	s.GetMutex().Lock()
+	defer s.GetMutex().Unlock()
+
+	ths := make([]*channel.Channel, len(g.Threads))
+	messages := make(map[string][]*channel.Message, len(g.Threads))
+	// converting channelIDs to map to have better perf
+	var channels map[string]string
+	if channelIDs != nil {
+		channels = make(map[string]string, len(channelIDs))
+		for _, id := range channelIDs {
+			channels[id] = id
+		}
+	}
+	// removing from map archived/deleted thread and saving untouched threads
+	i := 0
+	for _, c := range s.channelMap {
+		if c.GuildID == guildID && c.IsThread() {
+			// if thread is in sync list
+			ok := true
+			if channels != nil {
+				_, ok = channels[c.ID]
+			}
```

Comment:
Since the code seems to have moved, I refresh the comment:

How could channels be nil?